Use the executor for the part that can be parallell when rebuilding i…

baldersheim opened the pull request:
…ndex on load.

@toregge PR

baldersheim committed:
Use the executor for the part that can be parallell when rebuilding index on load.

toregge commented on searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp:
```diff
@@ -220,23 +223,45 @@ DenseTensorAttribute::onLoad(vespalib::Executor *executor)
-                if ((lid % 256) == 0) {
-                    commit();
+                if (executor != nullptr) {
+                    while (pending > 1000) { std::this_thread::yield(); }
```
Just yield causes wasted cpu usage on systems with many CPU cores. Consider using a condition variable.

baldersheim replied:
Refactored and fixed. PTAL

baldersheim committed:
Refactor for readability and maintenance.

toregge commented on searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp:
```diff
@@ -204,11 +204,84 @@ DenseTensorAttribute::extract_cells_ref(DocId docId) const
+                _cond.wait(guard);
+            }
+        }
+        while (_pending > 1000) { std::this_thread::yield(); }
```
Leftover code that should be removed ?

baldersheim replied:
Fixed

toregge commented on searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp:
```diff
@@ -204,11 +204,84 @@ DenseTensorAttribute::extract_cells_ref(DocId docId) const
+        ++_pending;
+        _shared_executor.execute(vespalib::makeLambdaTask([this, ref, lid]() {
+            auto prepared = _attr._index->prepare_add_document(lid, _attr._denseTensorStore.get_typed_cells(ref), _attr.getGenerationHandler().takeGuard());
+            _complete_executor.execute(vespalib::makeLambdaTask([this, lid, result=std::move(prepared)]() mutable {
```
Note that this allows "out of order" completions. If we want "in order" completions (as during feeding) then the shared executor should populate a promise and the complete executor task should take a future as argument and be scheduled in the foreground (not by the shared executor task).

baldersheim replied:
Out of order is not a problem, as we do not have any ordering guarantees in vespa either. As ANN is in nature approximative that should be even less of an issue.
However I rewrote to use a queue and let the foreground thread do the completion too.
That avoids creating an extra thread and any issues with 2 threads modifying the attribute, even if they did seem to be disjunct.

baldersheim committed:
Notify when _pending reaches zero.

baldersheim committed:
Avoid starting a separate thread for completing index insert.
Use a queue and do completition in the forground. That ensures only a single thread
modifying the attribute.

baldersheim commented:
@toregge PTAL

toregge commented on searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp:
```diff
@@ -213,57 +213,96 @@ class Loader {
-            _cond.wait(guard);
+
+    void complete(uint32_t lid, std::unique_ptr<PrepareResult> prepared) {
+        _attr.setCommittedDocIdLimit(std::max(_attr.getCommittedDocIdLimit(), lid + 1));
```
This assumes in-order completion. The documents with highest lid might not be visible after load due to a later
complete of a document with a lower lid when using out-of-order completions.
